File: employee/tests/end_to_end_test/features/steps/auth_steps.py
from behave import given, when, then
from tests.end_to_end_test.pages.login_page import LoginPage


@given(u'the application is running')
def step_impl(context):
    assert context.driver is not None


@given(u'the test database is already seeded with users')
def step_impl(context):
    assert context.db is not None


@given(u'the employee is on the login screen')
def step_impl(context):
    context.login_page = LoginPage(context.driver)
    context.login_page.open()


@when(u'the employee enters username "{username}"')
def step_impl(context, username):
    context.login_page.enter_username(username)


@when(u'the employee enters password "{password}"')
def step_impl(context, password):
    context.login_page.enter_password(password)


@when(u'the employee clicks the login button')
def step_impl(context):
    context.dashboard_page = context.login_page.click_login_button()


@then(u'the employee sees the auth message: "{flash_message}"')
def step_impl(context, flash_message):
    message = context.login_page.wait_for_element(
        context.login_page.LOGIN_MESSAGE
    ).text

    assert flash_message in message


@then(u'the employee is redirected to the employee dashboard')
def step_impl(context):
    context.dashboard_page.wait_for_url_contains("/app")
    assert "/app" in context.login_page.get_current_url()

@then(u'the employee is not redirected to the dashboard')
def step_impl(context):
    assert "/app" not in context.login_page.get_current_url()

@when(u'the employee does not input any value for username')
def step_impl(context):
    username_input = context.login_page.wait_for_element(
        context.login_page.USERNAME_FIELD
    )
    username_input.clear()


@then(u'the username field is selected')
def step_impl(context):
    active_element = context.driver.switch_to.active_element

    username_input = context.login_page.wait_for_element(
        context.login_page.USERNAME_FIELD
    )

    assert active_element == username_input


@given(u'the employee enters username "{username}"')
def step_impl(context, username):
    context.login_page.enter_username(username)


@when(u'the employee does not input any value for the password')
def step_impl(context):
    password_input = context.login_page.wait_for_element(
        context.login_page.PASSWORD_FIELD
    )
    password_input.clear()


@then(u'the password field is selected')
def step_impl(context):
    active_element = context.driver.switch_to.active_element

    password_input = context.login_page.wait_for_element(
        context.login_page.PASSWORD_FIELD
    )

    assert active_element == password_input


@given(u'the employee is logged in')
def step_impl(context):
    context.login_page = LoginPage(context.driver)
    context.login_page.open()

    context.login_page.enter_username("employee1")
    context.login_page.enter_password("password123")
    context.dashboard_page = context.login_page.click_login_button()
    # Waiting for this element was the only thing that stopped tests for being flakey.
    # Most likely due to the fact that button eventlisteners attached after fetch and dom is set up.
    # The button only does something after the page fetches all expenses. Otherwise, it is just html.
    context.dashboard_page.wait_for_element(
        context.dashboard_page.TABLE
    )
    assert "/app" in context.login_page.get_current_url()


@when(u'the employee clicks the logout button')
def step_impl(context):
    context.login_page = context.dashboard_page.click_logout_button()


@then(u'the employee is redirected to the login page')
def step_impl(context):
    context.login_page.wait_for_url_contains("/login")
    assert "/login" in context.login_page.get_current_url()

File: manager/src/test/java/com/revature/end_to_end_tests/steps/AuthSteps.java
package com.revature.end_to_end_tests.steps;

import com.revature.end_to_end_tests.context.TestContext;
import com.revature.end_to_end_tests.pages.DashboardPage;
import com.revature.end_to_end_tests.pages.LoginPage;
import io.cucumber.java.en.And;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import static org.junit.jupiter.api.Assertions.*;

public class AuthSteps {
    private final TestContext context;
    private LoginPage loginPage;
    private DashboardPage dashboardPage;

    public AuthSteps(TestContext context){
        this.context = context;
    }

    @Given("the application is running")
    public void theApplicationIsRunning() {
        assertNotNull(context.getDriver(), "WebDriver should be initialized");
    }

    @And("the test database is already seeded with users")
    public void theTestDatabaseIsAlreadySeededWithUsers() {
        // Write code here that turns the phrase above into concrete actions
        assertTrue(true);
    }

    @Given("the manager is on the login screen")
    public void theManagerIsOnTheLoginScreen() {
        // Write code here that turns the phrase above into concrete actions
        loginPage = context.getLoginPage();
        assertNotNull(loginPage, "LoginPage should not be null");
    }

    @When("the manager enters username {string}")
    public void theManagerEntersUsername(String username) {
        // Write code here that turns the phrase above into concrete actions
        loginPage.enterUsername(username);
    }

    @And("the manager enters password {string}")
    public void theManagerEntersPassword(String password) {
        // Write code here that turns the phrase above into concrete actions
        loginPage.enterPassword(password);
    }

    @And("the manager clicks the login button")
    public void theManagerClicksTheLoginButton() {
        // Write code here that turns the phrase above into concrete actions
        dashboardPage = loginPage.clickLogin();
    }

    @Then("the manager sees the message: {string}")
    public void theManagerSeesTheMessage(String arg0) {
        // Write code here that turns the phrase above into concrete actions
        WebElement message = loginPage.waitForElement(By.id("login-message"));
        assertTrue(message.getText().contains(arg0));
    }

    @And("the manager is redirected to the manager dashboard")
    public void theManagerIsRedirectedToTheManagerDashboard() {
        // Write code here that turns the phrase above into concrete actions
        WebElement title = dashboardPage.waitForElement(
                By.xpath("//h1[normalize-space()='Manager Expense Dashboard']"));

        assertTrue(title.getText().contains("Manager Expense Dashboard"));
    }

    @Then("the manager is not redirected to the dashboard")
    public void theManagerIsNotRedirectedToTheDashboard() {
        // Write code here that turns the phrase above into concrete actions
        assertNotEquals("Manager Expense Dashboard", context.getDriver().getTitle());
    }

    @When("the manager does not input any value for username")
    public void theManagerDoesNotInputAnyValueForUsername() {
        // Write code here that turns the phrase above into concrete actions
        loginPage.enterUsername("");
        loginPage.enterPassword("password123");
    }

    @Then("the username field is selected")
    public void theUsernameFieldIsSelected() {
        // Write code here that turns the phrase above into concrete actions
        WebElement active = context.getDriver().switchTo().activeElement();
        WebElement username = loginPage.waitForElement(By.id("username"));

        assertEquals(username, active);
    }

    @When("the manager does not input any value for the password")
    public void theManagerDoesNotInputAnyValueForThePassword() {
        // Write code here that turns the phrase above into concrete actions
        loginPage.enterUsername("manager1");
        loginPage.enterPassword("");
    }

    @Then("the password field is selected")
    public void thePasswordFieldIsSelected() {
        // Write code here that turns the phrase above into concrete actions
        WebElement active = context.getDriver().switchTo().activeElement();
        WebElement password = loginPage.waitForElement(By.id("password"));

        assertEquals(password, active);
    }

    @Given("the manager is logged in")
    public void theManagerIsLoggedIn() {
        // Write code here that turns the phrase above into concrete actions
        loginPage = context.getLoginPage();
        dashboardPage = loginPage.login("manager1","password123");
    }

    @When("the manager clicks the logout button")
    public void theManagerClicksTheLogoutButton() {
        // Write code here that turns the phrase above into concrete actions
        dashboardPage.logout();
    }

    @Then("the manager is redirected to the login page")
    public void theManagerIsRedirectedToTheLoginPage() {
        // Write code here that turns the phrase above into concrete actions
        assertEquals("Manager Login - Expense Manager", context.getDriver().getTitle());
    }
}
